### ISSUE TITLE
Flat target object, nested source objects, target = "." not working

### ISSUE THREAD

**Issue by afdia** [mapstruct/mapstruct#1406]
I often come across situations where objects have fields on different levels (e.g. target has flat fields, source has multiple nested objects and fields are on different levels)

Let's take the following example (getters/setters removed for clarity):
```java
public class MapstructExample {
public static class FlatTarget {
	public String targetA;
	public String targetB;
}
public static class NestedOuter {
	public String sourceA;
	public NestedInner nestedInner;
}
public static class NestedInner {
	public String sourceB;
}
@Mapper
public static interface MyMapper {
	@Mapping(target = "targetA", source = "sourceA")
	@Mapping(target = "targetB", source = "nestedInner.sourceB") // redundant mapping info
	//@Mapping(target = ".", source = "nestedInner") // not working
	FlatTarget convert1(NestedOuter in);

	@Mapping(target = "targetB", source = "sourceB") // redundant mapping info
	FlatTarget convert2(NestedInner in);

	@Mapping(target = "sourceA", source = "targetA")
	@Mapping(target = "nestedInner", source = ".") // no redundancy because "." works
	NestedOuter convert3(FlatTarget in);

	@Mapping(target = "sourceB", source = "targetB")
	NestedInner convert4(FlatTarget in);
}
}
```

As you can see the last 2 mappings have no repetition, because I can declare `source = "."` and it passes the main object to the next mapping method.
Unfortunately `target = "."` doesn't seem to work (and applying `@InheritInverseConfiguration(name="convert3")` to the first method causes and error)
I have also tried to formulate all converter method using `@MappingTarget` but it doesn't change the behavior.

Is there a way to avoid repeating the mappings of all fields of the NestedInner object? `@InheritConfiguration` is also not an option because the mappings are declared on a different layer (every field is prefixed with `nestedInner` in the second method).
Typically I work around the issue by declaring `@AfterMapping` methods which call the other method to fill in the missing parts of the objects, but I hope there is a better way to declare these mappings (or perhaps there is a way to get `target = "."` to work which would allow automatic inversion of mappings using `source = "."`)

------

EDIT: I have reformatted the example in order to have a fully working example and named the convert methods from 1-4 for clarity

**Comment by sjaakd** [comment]
Try to specify: `in.sourceA` and `in.nestedInner.sourceB`.. So include the parameter name as well.

**Comment by afdia** [comment]
Thanks for the suggestion. `source="."` and `source="in"` both work as expected, but my problem is that the target attribute doesn't work unless I also specify a property.

For example, the following mappings are working as expected. They are the same as in the initial code example, but I have changed them to use parameter names wherever possible. I also use `@MappingTarget` to have a parameter name for the output object.
```java
@Mapping(target = "out.targetA", source = "in.sourceA")
@Mapping(target = "out.targetB", source = "in.nestedInner.sourceB")
void convert(@MappingTarget FlatTarget out, NestedOuter in);

@Mapping(target = "targetA", ignore = true)
@Mapping(target = "out.targetB", source = "in.sourceB")
void convert(@MappingTarget FlatTarget out, NestedInner in);

@Mapping(target = "out.sourceA", source = "in.targetA")
@Mapping(target = "out.nestedInner", source = "in")
void convertOuter(@MappingTarget NestedOuter out, FlatTarget in);

@Mapping(target = "out.sourceB", source = "in.targetB")
void convertInner(@MappingTarget NestedInner out, FlatTarget in);
```
If I change the first method to the following code, in order to tell mapstruct "use the second method to map the subobject nestedInner", it shows an error `Unknown property "out" in result type at.soroban_it.soroservice.zmr.converter.MapstructExample.FlatTarget. Did you mean "targetA"?`
```java
@Mapping(target = "out.targetA", source = "in.sourceA")
@Mapping(target = "out", source = "in.nestedInner")
void convert(@MappingTarget FlatTarget out, NestedOuter in);
```

So using the target object itself (and not just one property of the target object) as parameter value of the `target=` attribute is currently not supported if I'm not mistaken (using `target="."` or `target="out"` doesn't make a difference here)

As mentioned before it's possible to get the expected behavior with an `@AfterMapping` method:
```java
@Mapping(target = "out.targetA", source = "in.sourceA")
//@Mapping(target = "out", source = "in.nestedInner") // this mapping IMO should do the same as the AfterMapping method
void convert(@MappingTarget FlatTarget out, NestedOuter in);

@AfterMapping
default void afterConvert(@MappingTarget FlatTarget out, NestedOuter in) {
	convert(out, in.nestedInner);
}
```
The workaround is quite simple and only 3 additional lines of code, but the main issue is that it's blocking me from using `@InheritInverseConfiguration` if the source mapping uses `source="."`, so I have to re-specify the inverse of many mappings which could theoretically be autogenerated by mapstruct.

-----

Btw one minor thing I have found while trying different solutions:
If I apply `@Mapping(target = "out.targetA", ignore = true)` it still shows `Unmapped target property: "targetA"`, so the ignore attribute doesn't seem to work when the parameter name is mentioned, but it's a very minor bug so I'm not sure if it's worth an issue

**Comment by sjaakd** [comment]
We designed this feature (nested source / nested target). It works on the premise that you include the full names (including the parameter). "." I'm pretty sure, we never tested :smile: . But, you can include the complete source-parameter. However, you cannot make a reverse mapping for such mappings.

Try to avoid mixd style. So defining your own methods and have the nested stuff, letting MapStruct automatically generate mappings.

The ignore of nested parameters should be working (and is part of our unit test). Reversing an ignored mapping does not work however. It is (I think) possible to include target-and-source-and-ignore to achieve that. But I'm not entirely sure.. We should document that..

Anyway... I'll have a more detailed look at your examples later this week. Perhaps you can try out some stuff as well with the above knowledge.

**Comment by sjaakd** [comment]
I'm still not sure about this one.. The method is a bean mapping method. Can you still speak of a bean mapping method, if the target itself is mapped? What would that mean? You can only point out one source property in that case?

**Comment by afdia** [comment]
> Can you still speak of a bean mapping method, if the target itself is mapped? What would that mean?

I'm not sure about the internals of MapStruct, so perhaps I misunderstand the question, but in my first example I wrote
```java
//@Mapping(target = ".", source = "nestedInner") // not working
```
to tell MapStruct "try to find a mapping from the parameter type of `nestedInner `(NestedInner.class) to the parameter type of `.` (FlatTarget.class) and I assumed that it will use the following mapping method:
```java
FlatTarget convert2(NestedInner in);
```
because the inverse mapping with `source="."` works. The difference I see to the inverse mapping is that the source is guaranteed to exist and the target must be instantiated before mapping, therefore the method convert2 cannot be directly used (for that it would have to use a `@MappingTarget`  to receive the partially filled target object and add only the mappings which are specified in the convert2 method)
So if the user doesn't provide the `@MappingTarget` variant of the mapping method, MapStruct would have to create and use an additional "companion method" which looks similar to the one generated by
```java
void convert2(@MappingTarget FlatTarget out, NestedInner in);
```

------

> You can only point out one source property in that case?

One possible usecase of Mapstruct could be splitting up fields of a large object to a target which has several nested objects, all of which contain parts of the large object.
So e.g. (untested but it should work):

```java
public static class Target {
	public NestedTargetA targetA;
	public NestedTargetB targetB;
}
public static class NestedTargetA {
	public String a;
}
public static class NestedTargetB {
	public String b;
}
public static class Source {
	public String a;
	public String b;
}
@Mapper
public static interface MyMapper {
	@Mapping(target = "targetA", source = ".")
	@Mapping(target = "targetB", source = ".")
	Target convert1(Source in);

	NestedTargetA convert2(Source in);

	NestedTargetB convert3(Source in);
}
```

If we try to invert the mappings, we would end up with multiple target="." mappings which have different sources, something like:

```java
@Mapper
public static interface MyMapper {
	@Mapping(source = "targetA", target = ".")
	@Mapping(source = "targetB", target = ".")
	Source convert1(Target in);

	Source convert2(NestedTargetA in);

	Source convert3(NestedTargetB in);
}
```

So I guess there is a usecase for multiple mappings which use `target="."` and I would assume that the implementation of convert1 creates the target object, then calls convert2, then convert3 on the created target object (as mentioned above using companion methods which use the `@MappingTarget` syntax to avoid that convert2 and convert3 create new Source objects)

Of course there is also the question of overlapping mappings. What if `NestedTargetA `and `NestedTargetB` contain a field with the same name which is also in `Source`? Then calling convert3 would overwrite the mapping from overwrite2, therefore the order of the applied mapping annotations would be relevant (the latest one wins).

But I guess this is a very specific case and "latest wins" is still a fine way to resolve the ordering issue.
In general I would love to have a working `target="." ` mapping, because many mappings could be inverted or just simplified when working with object structures which use "different layers of nesting"

**Comment by lizhecao** [comment]
+1

**Comment by sjaakd** [comment]
fixed by PR #1833 and #1920
